orchestrator-salt/src/main/java/com/sequenceiq/cloudbreak/orchestrator/salt/SaltOrchestrator.java

```diff
@@ -782,20 +783,38 @@ public void stopClusterManagerAgent(GatewayConfig gatewayConfig, Set<Node> nodes
     public void leaveDomain(GatewayConfig gatewayConfig, Set<Node> allNodes, String roleToRemove, String roleToAdd, ExitCriteriaModel exitCriteriaModel)
             throws CloudbreakOrchestratorFailedException {
         try (SaltConnector sc = createSaltConnector(gatewayConfig)) {
-            Set<String> targetHostnames = allNodes.stream().map(Node::getHostname).collect(Collectors.toSet());
-            saltCommandRunner.runSaltCommand(sc, new GrainAddRunner(targetHostnames, allNodes, "roles", roleToAdd), exitCriteriaModel, maxRetryLeave,
-                    exitCriteria);
-            saltCommandRunner.runSaltCommand(sc, new GrainRemoveRunner(targetHostnames, allNodes, "roles", roleToRemove), exitCriteriaModel, maxRetryLeave,
-                    exitCriteria);
-            Set<String> allHostnames = allNodes.stream().map(Node::getHostname).collect(Collectors.toSet());
-            saltCommandRunner.runSaltCommand(sc, new SyncAllRunner(allHostnames, allNodes), exitCriteriaModel, maxRetryLeave, exitCriteria);
-            runNewService(sc, new HighStateRunner(allHostnames, allNodes), exitCriteriaModel, maxRetryLeave, true);
+            if (isChangingRolesNecessary(gatewayConfig, sc, roleToRemove)) {
+                Set<String> targetHostnames = allNodes.stream().map(Node::getHostname).collect(Collectors.toSet());
+                saltCommandRunner.runSaltCommand(sc, new GrainAddRunner(targetHostnames, allNodes, "roles", roleToAdd),
+                        exitCriteriaModel, maxRetryLeave, exitCriteria);
+                saltCommandRunner.runSaltCommand(sc, new GrainRemoveRunner(targetHostnames, allNodes, "roles", roleToRemove),
+                        exitCriteriaModel, maxRetryLeave, exitCriteria);
+                Set<String> allHostnames = allNodes.stream().map(Node::getHostname).collect(Collectors.toSet());
+                saltCommandRunner.runSaltCommand(sc, new SyncAllRunner(allHostnames, allNodes), exitCriteriaModel, maxRetryLeave, exitCriteria);
+                runNewService(sc, new HighStateRunner(allHostnames, allNodes), exitCriteriaModel, maxRetryLeave, true);
+            }
         } catch (Exception e) {
             LOGGER.info("Error occurred during executing highstate (for recipes).", e);
             throw new CloudbreakOrchestratorFailedException(e);
         }
     }
 
+    private boolean isChangingRolesNecessary(GatewayConfig gatewayConfig, SaltConnector sc, String role) {
+        return getMemberRoles(gatewayConfig, sc).contains(role);
+    }
+
+    private Set<String> getMemberRoles(GatewayConfig gatewayConfig, SaltConnector sc) {
+        Map<String, JsonNode> roles = SaltStates.getGrains(sc, new HostList(List.of(gatewayConfig.getHostname())), "roles");
+        return roles.entrySet().stream().findFirst()
+                .map(Entry::getValue)
```

Comment:
why do you stream the entryset if you need only the values?

Reply:
ok